src/app/model/ChessEngine.ts

```diff
@@ -1,3 +1,50 @@
 import { ChessModel } from './ChessModel';
+import { Side, Square } from './Types';
+import { Piece } from './Piece';
+import { Player } from './Player';
+import { ChessBoard } from './ChessBoard';
+import { PieceWasMoved } from './PieceWasMoved';
+import { PieceWasCaptured } from './PieceWasCaptured';
 
-export class ChessEngine implements ChessModel {}
+export class ChessEngine implements ChessModel {
+  currentSide: Side | undefined;
+
+  constructor(readonly board: ChessBoard) {}
+
+  move(byPlayer: Player, squareFrom: Square, squareTo: Square): (PieceWasMoved | PieceWasCaptured)[] {
+    const chosenPiece = this.board.onPositionPiece(squareFrom) as Piece; // tutaj chyba też powinienem pozbyć się tego as Piece
+    const possibleMoves = chosenPiece.possibleMoves(squareFrom, this.board);
+    const pieceOnSquare = this.board.onPositionPiece(squareTo);
+    const isOponent = pieceOnSquare && pieceOnSquare.side !== chosenPiece.side;
+
+    if (byPlayer.side == this.currentSide) throw new Error('balalab');
+    if (byPlayer.side !== chosenPiece.side) throw new Error('Player can not move other players pieces.');
+    if (
+      !possibleMoves.find((square: Square) => {
+        return square.column == squareTo.column && square.row == squareTo.row;
+      })
+    )
+      throw new Error('Piece can not move to given square.');
+    const pieceWasMoved: PieceWasMoved = { eventType: 'PieceWasMoved', piece: chosenPiece, from: squareFrom, to: squareTo };
+    this.onPieceWasMoved(pieceWasMoved);
+    const pieceWasCaptured: PieceWasCaptured | undefined =
+      isOponent && isPiece(pieceOnSquare)
+        ? {
+            //dodałem unię z undefined bo mi krzyczy
+            eventType: 'PieceWasCaptured',
+            piece: pieceOnSquare,
+            onSquare: squareTo,
+          }
+        : undefined;
+
+    return pieceWasCaptured ? [pieceWasMoved, pieceWasCaptured] : [pieceWasMoved];
+  }
+
+  private onPieceWasMoved(event: PieceWasMoved): void {
+    this.board.movePiece(event.from, event.to);
+    this.currentSide = event.piece.side;
+  }
+}
+
+//Type Guard for Piece - wyczytałem, że można stworzyć taką funkcję 'pilnującą' albo możemy skorzystać z instanceof
+const isPiece = (varToCheck: unknown): varToCheck is Piece => (varToCheck as Piece) !== undefined;
```

Comment:
Przepraszam bardzo, ale ten... Type Guard jest to porstu głupi :P 
Tzn. on ma sens w przpadku np:
```
const isPieceWasMoved = (event: unknown): event is PieceWasMoved => event?.eventType === 'PieceWasMoved';
```
Jak mamy klasę Piece to jak najbardziej robimy instance of. 
Ale w tym przypadku to wystarczy nam bardziej isDefined. A to można zrobić generycznie:

Takie 2 funkcje czesto sie przydaja w projekcie:
```
export function isDefined<T>(x: T | undefined): x is T {
    return typeof x !== "undefined" && x !== null;
}

export function isNotDefined<T>(x: T | undefined): x is undefined {
    return !isDefined(x);
}
```
mozna to do jakiejs wspolnego pliku dac :)

Reply:
To as Piece psuje całe działanie.
`(varToCheck as Piece)`
Spróbuj tam przekazać coś co nie bedzie piece.